### PR TITLE
bitstamp fetchTransactions fixes

### PR DESCRIPTION
**Pull request by firepol**
This fixes 3 issues:

1) `let result = this.filterByArray (response, 'type', [ '0', '1' ]);` doesn't work as expected (at least in python) and I invite you to test it further, as result I was getting my first bitstamp withdrawal. All the rest was missing.

2) Withdrawals were not parsed correctly (currency and amount: undefined) because of  `if ((value !== undefined) && (value > 0))` in the `getCurrencyIdFromTransaction` method. Withdrawls have an amount < 0, so that code was skipped for all withdrawals.

3) if you pass a currency to fetch transactions only for that currency, the filtering by that currency was not done. Now it is. Not sure if you have a better way to _clean_ the currency. I know it must be uppercase, so that's the only thing I did. Maybe there is a better way.

### CODE REVIEW

**Comment by kroitor** [comment]
@firepol hey) Thx again for the contribution! ) Can I ask you to paste the `verbose` structures for those transactions that weren't parsed properly, plz, so that I could debug `filterByArray` as well?

**Comment by firepol** [comment]
Hi @kroitor, I think there is no need to send you verbose data for the structures not parsed, the explanation is very easy.

The issue is in the method `getCurrencyIdFromTransaction`, in this loop, which I comment here line by line, because at first sight in the morning I also didn't immediately understand it:

```
// ids: all keys of the transaction object
let ids = Object.keys (transaction);
for (let i = 0; i < ids.length; i++) {
    let id = ids[i];
    // if a key has `_` it means it's a pair, e.g. btc_usd, we skip it
    if (id.indexOf ('_') < 0) {
        let value = this.safeFloat (transaction, id);
        // value is the amount, e.g. `'btc': 0.15` it means the transaction was 0.15 BTC
        // bitstamps list all the other currencies with amount = 0.0, e.g. `'xrp': 0.0`
        // if the transaction is a withdrawal, the amount is negative, e.g. `'btc': -0.1`
        if ((value !== undefined) && (value !== 0)) {
            // if the amount is 0, we can ignore it, else we return it, as it's the currency used for this transaction
            return id;
        }
    }
}
// Previously, `undefined` was returned for transactions which amount was negative, not, as I fixed it, the currencyId is returned properly and the `parsetransactions` method works properly.
return undefined;
```

In `parseTransaction` note this call `currencyId = this.getCurrencyIdFromTransaction (transaction)`; 

The currencyId was set to `undefined` for all withdrawals, as their amount was negative. So the amount was not set, currency not set etc.

Basically the structure you want to see will look like this (note `"eth": "-0.05000000",` the amount is negative, as it's negative, the parsing is skipped and everything stays `undefined`):

```
{
   "fee": "0.00000000",
   "btc_usd": "0.00",
   "datetime": XXX,
   "usd": 0.0,
   "btc": 0.0,
   "eth": "-0.05000000",
   "type": "0",
   "id": XXX,
   "eur": 0.0
}
```

Hope it's clear now.

**Comment by kroitor** [comment]
I wanted to see the actual types of the values they used, as this is not clear from their docs, namely the type of the `type` field – is it actually a string in quotes or is it an integer? The value of eth in your example is in quotes, other values are zeroed floats, so, you did copy the structure that was mocked in the comments in my code already, however I would not ask you of actual `verbose` structures in the first place, if that was enough (because I know it's there already – it does not answer my questions). I don't really need to see the values, so, you can edit them if you don't want to show them, I'm mostly interested in the actual types of those fields. So, if you could post some of your output, that would still be very helpful for my needs (that is, apart from the issue in this PR).

**Comment by firepol** [comment]
Ok, I will share some verbose outputs if that helps you.

For sure now I can already tell you:

- `type`: string, e.g. `"0"`
- `amount`: float e.g.: `0.0`

But I will post one example of deposit, one withdrawal, one trade. I read the manual and saw "for python" people I'll add this in my code:

```
import logging
logging.basicConfig(level=logging.DEBUG)
```

**Comment by firepol** [comment]
Hi @kroitor 

here some data for you (oh now I see the amount is not always a string, for the currencies that don't matter, the zero is actually an int...), hope this helps. If you need more examples let me know.

```
[
  {
    "fee": "0.06",
    "order_id": 9994567896,
    "id": 1234567896,
    "usd": 0,
    "xrp_eur": 0.2257,
    "btc": 0,
    "datetime": "2018-09-10 16:48:55",
    "type": "2",
    "xrp": "100.00000000",
    "eur": "-22.570000"
  },
  {
    "fee": "0.03",
    "order_id": 9994567895,
    "datetime": "2018-09-09 07:00:22",
    "usd": 0,
    "btc": 0,
    "eth": "0.07000000",
    "eth_eur": 165.72,
    "type": "2",
    "id": 1234567895,
    "eur": "-11.600"
  },
  {
    "fee": "0.00000000",
    "btc_usd": "0.00",
    "id": 1234567894,
    "usd": 0,
    "btc": 0,
    "datetime": "2018-09-08 09:00:31",
    "type": "1",
    "xrp": "-20.00000000",
    "eur": 0
  },
  {
    "fee": "0.28000000",
    "order_id": 9994567893,
    "datetime": "2018-09-07 13:01:44",
    "usd": 0,
    "btc": "-0.02000000",
    "btc_eur": 5553,
    "type": "2",
    "id": 1234567893,
    "eur": "111.06"
  },
  {
    "fee": "0.28",
    "order_id": 9994567892,
    "datetime": "2018-09-07 10:11:33",
    "usd": 0,
    "btc": "0.02000000",
    "btc_eur": 5487,
    "type": "2",
    "id": 1234567892,
    "eur": "-109.74"
  },
  {
    "fee": "0.00000000",
    "btc_usd": "0.00",
    "datetime": "2018-01-10 14:11:22",
    "usd": 0,
    "btc": "-0.01500000",
    "type": "1",
    "id": 1234567891,
    "eur": 0
  },
  {
    "fee": "0.90",
    "btc_usd": "0.00",
    "datetime": "2017-12-15 13:00:00",
    "usd": 0,
    "btc": 0,
    "type": "1",
    "id": 1234567890,
    "eur": "-19.10"
  }
]
```

**Comment by firepol** [comment]
@kroitor not at all, thanks to you for maintaining this awesome project, for the fast and kind replies and fast (compared to many other open source projects where they are either ignored or forgotten for weeks, months) PR merges and feedbacks. I appreciate this and I like contributing to this project.
